l1-contracts/contracts/bridge/asset-tracker/L1AssetTracker.sol

```diff
@@ -0,0 +1,369 @@
+            /// For chains that were settling on GW before V30, we need to update the chain's chainBalance until the chain updates to V30.
+            chainToUpdate = settlementLayer == 0 || l2BatchNumber < v30UpgradeChainBatchNumber
+                ? _chainId
+                : settlementLayer;
+        } else {
+            chainToUpdate = settlementLayer == 0 ? _chainId : settlementLayer;
+        }
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                    Gateway related token balance migration 
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice This function receives the migration from the L2 or the Gateway.
+    /// @dev It sends the corresponding L1->L2 messages to the L2 and the Gateway.
+    function receiveMigrationOnL1(FinalizeL1DepositParams calldata _finalizeWithdrawalParams) external {
+        _proveMessageInclusion(_finalizeWithdrawalParams);
+
+        (bytes4 functionSignature, TokenBalanceMigrationData memory data) = DataEncoding
+            .decodeTokenBalanceMigrationData(_finalizeWithdrawalParams.message);
+        require(
+            functionSignature == IAssetTrackerDataEncoding.receiveMigrationOnL1.selector,
+            InvalidFunctionSignature(functionSignature)
+        );
+
+        require(assetMigrationNumber[data.chainId][data.assetId] < data.migrationNumber, InvalidAssetId(data.assetId));
+
+        uint256 currentSettlementLayer = _bridgehub().settlementLayer(data.chainId);
+        uint256 fromChainId;
+        uint256 toChainId;
+
+        if (data.isL1ToGateway) {
+            uint256 chainMigrationNumber = _getChainMigrationNumber(data.chainId);
+            /// We check the chainMigrationNumber to make sure the message is from a previous token migration.
+            require(
+                chainMigrationNumber == data.migrationNumber,
+                InvalidChainMigrationNumber(chainMigrationNumber, data.migrationNumber)
+            );
+            /// In this case the TokenBalanceMigrationData data might be malicious.
+            /// We check the chainId to match the finalizeWithdrawalParams.chainId.
+            /// We check the assetId, tokenOriginChainId, originToken with an assetIdCheck.
+            /// The amount might be malicious, but that poses a restriction on users of the chain, not other chains.
+            /// The AssetTracker cannot protect individual users only other chains. Individual users rely on the proof system.
+            /// The last field is migrationNumber, which cannot be abused.
+
+            require(currentSettlementLayer != block.chainid, NotMigratedChain());
+            require(data.chainId == _finalizeWithdrawalParams.chainId, InvalidWithdrawalChainId());
+
+            // we check parity here to make sure that we migrated the token balance back to L1 from Gateway.
+            // this is needed to ensure that the chainBalance on the Gateway AssetTracker is currently 0.
```

Comment:
Above we require that `chainMigrationNumber == data.migrationNumber` and `currentSettlementLayer != block.chainid`. From the above it follows that `chainMigrationNumber` is odd.

The below parity checks just checks that the parity of the asset migration number is even. Does it make sense to just replace it with simple `require(assetMigrationNumber % 2 == 0)`?

If you are afraid of having "`chainMigrationNumber` is odd." as an implicit requirement, maybe it makes sense to just make a separate require for it.